docs/slides/day3.md

```diff
@@ -0,0 +1,272 @@
+
+---
+
+# Software Development Process
+
+![Software Dev Process meme](day3_images/software-dev-process.jpg)
+
+---
+
+# Warmup
+
+Share what you learned from our last session about conditionals (if, else if, else…)
+
+Have a classmate code review your election time code (and vice versa).
+
+* Is it clear what the program is doing?
+* Are there sufficient comments?
+* Does the control flow of the program make sense?
+* Does the code follow the ruby style guide?
+* Is there repeated code anywhere?
+* Are edge cases appropriately handled?
+
+Focus on using programming vocabulary (e.g, blocks, loops, iterator, variable assignment, conditional statement) and be as precise as possible when reviewing.
+
+---
+
+# Loops
+
+Code that can be repeated more than once
+
+What's the difference between the following types of loops?
+
+* Sentinel Controlled
+* Counter Controlled
+* Infinite
+
+
+---
+
+# Sentinel Controlled
+
+.left-column[
+    While Loop
+
+```ruby
+while <something is true>
+  # do stuff
+end
+```
+]
+
+.right-column[
+    Until Loop
+
+```ruby
+until <something is false>
+  # do stuff
+end
+```
+]
+
+--
+
+## When would you use an `until` instead of a `while`?
+
+---
+
+# Convert to an until loop
+
+.left-column[
+    ```ruby
+    puts "Hello we're going to total 
+      some numbers!"
+    puts "Enter a negative number 
+      to quit."
+
+    total = 0
+    input = gets.chomp.to_i
+
+    while input > -1
+      total += input
+      input = gets.chomp.to_i
+    end
+
+    puts "Total is #{total}"
+    ```
+]
+
+.right-column[
+    ```ruby
+    puts "Hello we're going to total 
+      some numbers!"
+    puts "Enter a negative number 
+      to quit."
+
+    total = 0
+    input = gets.chomp.to_i
+
+    until input <= -1
+      total += input
+      input = gets.chomp.to_i
+    end
+
+    puts "Total is #{total}"
+    ```
+]
+
+---
+
+# Counter Controlled
+
+## Times Loop
+
+```ruby
+<Integer>.times do |i|  # i is optional
+  # code goes here
+end
+```
+
+--
+
+When do you use such a loop?
+
+---
+
+# What kind of loop is this?
+
+```ruby
+print "Do you like dogs or cats?"
+pet = gets.chomp.downcase
+
+until pet == "dogs"
+  puts "How rude!  I don't believe you, how can anyone not prefer dogs!"
+  print "Come now, tell me truthfully, do you like cats or dogs?"
+end
+
+puts "Oh you like dogs now too!  We'll be such good friends!"
+```
+
+--
+
+## Infinite!
+
+---
+
+# Iterators
+
+Useful for looping over a collection
+
+.left-column[
+```ruby
+  numbers = [1, 2, 3, 4, 5]
+  numbers.each do |number|
+    puts number
+  end
+```
+]
+
+.right-column[
+  Iterator
+
+  ```ruby
+  numbers = [1, 2, 3, 4, 5]
+*numbers.each do |number|
+   puts number
+  end
+```
+
+  Block
+  ```ruby
+    numbers = [1, 2, 3, 4, 5]
+  numbers.each do |number|
+*   puts number
+*  end
+  ```
+]
+
+---
+
+# What kind of loop is this?
+
+.left-column[
+    ```ruby
+    numbers = [1, 2, 3, 4, 5]
+    numbers.each do |number|
+      puts number
+    end
+    ```
+]
+
```

Comment:
Same comment here - if you meant to have a pause between showing the left and then right column, add a `--` line between